storage/innobase/srv/srv0srv.cc

```diff
@@ -1750,14 +1837,42 @@ static uint32_t srv_do_purge(ulint* n_total_purged)
 			std::lock_guard<std::mutex> lk(purge_thread_count_mtx);
 			n_threads = n_use_threads = srv_n_purge_threads;
 			srv_purge_thread_count_changed = 0;
-		} else if (trx_sys.history_size_approx() > rseg_history_len
-			   || (srv_max_purge_lag > 0
-			       && rseg_history_len > srv_max_purge_lag)) {
+
+			compute_series(series, innodb_purge_threads_MAX + 1,
+				(100 - (adaptive_purge_threshold + safety_net)),
+				n_threads);
+			lsn_lwm = adaptive_purge_threshold;
+			lsn_hwm = adaptive_purge_threshold + series[n_threads];
+			start_time = 0;
+			lsn_age_factor = get_lsn_age_factor();
+		} else if (trx_sys.history_size_approx() > rseg_history_len) {
+
+			/* dynamically adjust the purge thread based on redo log
+			fill factor */
+			if (n_use_threads < n_threads &&
+				lsn_age_factor < lsn_lwm) {
+				++n_use_threads;
+				lsn_hwm = lsn_lwm;
+				lsn_lwm -= series[n_use_threads];
+			} else if (n_use_threads > 1 &&
+				lsn_age_factor >= lsn_hwm) {
+				--n_use_threads;
+				lsn_lwm = lsn_hwm;
+				lsn_hwm += series[n_use_threads];
+			} else if (n_use_threads == 1 &&
+				lsn_age_factor >= (100 - safety_net)) {
+				std::this_thread::sleep_for(
+					std::chrono::milliseconds(10));
```

Comment:
Instead of sleeping in the purge coordinator callback function, could we please invoke `purge_coordinator_timer->set_time()` to schedule the next call?

Reply:
as I understand, purge_coordinator_timer will delay the scheduling of the next coordinator by 10 units (10 ms) but once the coordinator has started it will continue to work till there is work to do and this is where we need active sleep.

Reply:
I think it might be a good idea generally to refactor this chunk into a "recommend_concurrency_level" pure function, that does some math, and  returns "increase", "decrease", or "delay", and another function that does "update_concurrency_level" that actually changes n_use_threads and hwm/lwm.   I'd be glad to see the "AI" logic separated into dedicated functions,  would make it easier for the reader to understand the logic.

The action to "delay" action could work by disabling the purge temporarily, and reenable it in a timer action, scheduled 10ms from now. This is only because blocking threadpool thread is "considered harmful", the thread could do something more useful that sleeping, also the pool assumes the thread is CPU-bound, and will throttle other tasks if all CPUs are busy. although I do not know  if it is such a big deal .

Reply:
1. suggestion on having a separate logic. yes, it could be explored for better readability if that helps.

2. ref: purge-sleep. <I am still trying to understand the threadpool so correct if anything is wrong>

a. are you suggesting we pause/stop the purge like we do for flush table for export and reenable it (resume) as part of the timer action purge_coordinator_timer_callback(). as of currently, purge_coordinator_timer_callback() will simply return if the purge is paused so if we need to resume it for the said use-case then the reason for the pause will have to capture.
but as you say how much effect this all would have vs active sleep of callback thread would decide if this is worth chasing.